### DISCUSSION TITLE
removal of if statements in _handle_consumable

### DISCUSSION THREAD

**Discussion by ghost** [SelinaDev/Godot-Roguelike-Tutorial#6] [Ideas]
Title, is that possible? i'm trying to make it so the script automatically handles any components it has, rather than making more and more if statements for every item i make, which scales badly

**Comment by SelinaDev** [comment]
I have thought about that as well, both in terms of how to handle arbitrary components and how to handle components with some variation, like the consumable component. I probably won't implement it in this version of the tutorial, as it would mean some refactoring, but a solution that came to my mind (and which I would probably use if I would redo this) is to have the `ConsumableComponentDefinition` class create the actual components, via a function.

Here's an example how that could work with one of the classes. _consumable_component_definition.gd_ would be expanded like so:
```
class_name ConsumableComponentDefinition
extends Resource

func reify() -> ConsumableComponent:
   return null
```

Then, in the concrete sub-classes you could implement the proper function. For example, _healing_consumable_component_definition.gd_ would look like this:

```
class_name HealingConsumableComponentDefinition
extends ConsumableComponentDefinition

@export var healing_amount: int = 0

func reify() -> ConsumableComponent:
   return HealingConsumableComponent.new(self)
```

You still would need to type all the different types in each of the definitions. However, I do feel like it would improve organization. Then, in the entity you could just use the following line:
```
if entity_definition.consumable_definition:
   consumable_component = entity_definition.consumable_definition.reify()
```
Each definition class would then be responsible for instantiating the correct type. And as you are likely to create consumable components and their definitions together you can just add or remove consumable components without the need to edit _entity.gd_. The only thing to remember is to implement the `reify()` function (or however you want to call it).

**Reply by ghost** [reply]
i forgot to write a reply, but thank you so much! i've implemented this a while ago based on your suggestions and made it work wonderfully.
